File: ios/Packages/FeatureServices/StreamService/StreamObserverService.swift
// Copyright (c). Gem Wallet. All rights reserved.

import Foundation
import protocol Gemstone.GemStreamServiceProtocol
import GemstonePrimitives
import Primitives
import WebSocketClient

public actor StreamObserverService: Sendable {
    private let service: any GemStreamServiceProtocol
    private let webSocket: any WebSocketConnectable
    private let health: ConnectionComponentHealth
    private let reconnection: any Reconnectable
    private var observeTask: Task<Void, Never>?
    private var isActive = false
    private var failedAttempts: UInt32 = 0
    private var connectedAt: ContinuousClock.Instant?

    public init(
        service: any GemStreamServiceProtocol,
        webSocket: any WebSocketConnectable,
        health: ConnectionComponentHealth,
        reconnection: any Reconnectable,
    ) {
        self.service = service
        self.webSocket = webSocket
        self.health = health
        self.reconnection = reconnection
    }

    deinit {
        observeTask?.cancel()
    }

    // MARK: - Public API

    public func connect() {
        isActive = true
        startObserving()
    }

    public func updateSession() async {
        startObserving()
        do {
            try await service.updateSession()
        } catch {
            debugLog("stream session update error: \(error)")
        }
    }

    public func disconnect() {
        isActive = false
        observeTask?.cancel()
    }

    // MARK: - Private

    private func startObserving() {
        guard isActive, observeTask == nil else { return }
        observeTask = Task { [weak self] in
            await self?.observeConnection()
        }
    }

    private func observeConnection() async {
        let failed: Bool
        do {
            try await runConnection()
            failed = false
        } catch is CancellationError {
            failed = false
        } catch {
            debugLog("stream connection error: \(error)")
            failed = true
        }
        health.report(isHealthy: false)
        await webSocket.disconnect()
        await service.disconnected()
        if failed {
            let next = reconnection.reconnection(attempt: failedAttempts, connectedFor: connectedAt.map { $0.duration(to: .now) } ?? .zero)
            failedAttempts = next.nextAttempt
            try? await Task.sleep(for: next.delay)
        }
        connectedAt = nil
        observeTask = nil
        if Task.isCancelled || failed {
            startObserving()
        }
    }

    private func runConnection() async throws {
        let shouldConnect = try await service.prepareConnection()
        try Task.checkCancellation()
        debugLog("stream connecting: \(shouldConnect)")
        guard shouldConnect else { return }
        for await event in await webSocket.connect() {
            try Task.checkCancellation()
            switch event {
            case .connected:
                debugLog("stream connected")
                try await service.connected()
                connectedAt = .now
                health.report(isHealthy: true)
            case let .message(data):
                await onMessage(data)
            case .disconnected:
                debugLog("stream disconnected")
                if isActive {
                    health.report(isHealthy: false)
                }
                await service.disconnected()
            }
        }
    }

    private func onMessage(_ data: Data) async {
        do {
            let event = try await service.decodeEvent(event: String(decoding: data, as: UTF8.self))
            debugLog("stream event: \(event)")
            Task { [service] in
                do {
                    try await service.sync(event: event)
                } catch {
                    debugLog("stream sync error: \(error)")
                }
            }
        } catch {
            debugLog("stream dropped an event: \(error)")
        }
    }
}

File: android/data/services/gemstone/src/main/kotlin/com/gemwallet/android/data/services/gemstone/stream/StreamObserverService.kt
package com.gemwallet.android.data.services.gemstone.stream

import android.util.Log
import com.gemwallet.android.application.session.cases.GetSession
import com.gemwallet.android.data.services.gemstone.connection.ConnectionComponentHealth
import com.gemwallet.android.ext.runCatchingCancellable
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.NonCancellable
import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.delay
import kotlinx.coroutines.ensureActive
import kotlinx.coroutines.flow.distinctUntilChanged
import kotlinx.coroutines.flow.drop
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import uniffi.gemstone.GemConnectionServiceInterface
import uniffi.gemstone.GemStreamServiceInterface
import kotlin.time.TimeMark
import kotlin.time.TimeSource

class StreamObserverService(
    private val getSession: GetSession,
    private val service: GemStreamServiceInterface,
    private val connection: WebSocketConnectable,
    private val health: ConnectionComponentHealth,
    private val connectionService: GemConnectionServiceInterface,
    private val scope: CoroutineScope = CoroutineScope(Dispatchers.IO),
) {
    private var connectionJob: Job? = null

    @Synchronized
    fun start() {
        if (connectionJob?.isActive == true) return
        val previousJob = connectionJob
        connectionJob = scope.launch {
            withContext(NonCancellable) { previousJob?.join() }
            currentCoroutineContext().ensureActive()
            val wallets = getSession().map { it?.wallet?.id?.id }.distinctUntilChanged()
            launch {
                wallets.drop(1).collect {
                    runCatchingCancellable { service.updateSession() }
                        .onFailure { Log.e(TAG, "Stream session update error", it) }
                }
            }
            var failedAttempts = 0u
            while (isActive) {
                var subscribedAt: TimeMark? = null
                runCatchingCancellable {
                    val connects = service.prepareConnection()
                    currentCoroutineContext().ensureActive()
                    when (connects) {
                        true -> observeConnection(onSubscribed = { subscribedAt = TimeSource.Monotonic.markNow() })
                        false -> wallets.first { it != null }
                    }
                }.onFailure {
                    Log.e(TAG, "Stream connection error", it)
                    val reconnection = connectionService.reconnection(failedAttempts, subscribedAt.connectedDuration())
                    failedAttempts = reconnection.nextAttempt
                    delay(reconnection.delay.toMillis())
                }
            }
        }
    }

    @Synchronized
    fun stop() {
        connectionJob?.cancel()
    }

    private suspend fun observeConnection(onSubscribed: () -> Unit) {
        try {
            connection.connect().collect { event ->
                when (event) {
                    WebSocketEvent.Connected -> {
                        service.connected()
                        health.report(isHealthy = true)
                        onSubscribed()
                    }

                    is WebSocketEvent.Message -> runCatchingCancellable {
                        val handled = service.decodeEvent(event.text)
                        scope.launch {
                            runCatchingCancellable { service.sync(handled) }
                                .onFailure { Log.e(TAG, "Stream sync error", it) }
                        }
                    }.onFailure { Log.e(TAG, "Stream event error", it) }

                    WebSocketEvent.Disconnected -> {
                        health.report(isHealthy = false)
                        runCatchingCancellable { service.disconnected() }
                            .onFailure { Log.e(TAG, "Stream event error", it) }
                    }
                }
            }
        } finally {
            health.report(isHealthy = false)
            withContext(NonCancellable) {
                runCatchingCancellable { service.disconnected() }
                    .onFailure { Log.e(TAG, "Stream disconnect error", it) }
            }
        }
    }

    companion object {
        private const val TAG = "StreamObserverService"
    }
}
